lib/service/service.go

```diff
@@ -5150,3 +5159,148 @@ func writeHostIDToKubeSecret(ctx context.Context, kubeBackend kubernetesBackend,
+		grpc.KeepaliveParams(keepalive.ServerParameters{
+			// Using an aggressive idle timeout here since this gRPC server
+			// currently only hosts the join service, which has no need for
+			// long-lived idle connections.
+			//
+			// The reason for introducing this is that teleport clients
+			// before #17685 is fixed will hold connections open
+			// indefinitely if they encounter an error during the joining
+			// process, and this seems like the best way for the server to
+			// forcibly close those connections.
+			//
+			// If another gRPC service is added here in the future, it
+			// should be alright to increase or remove this idle timeout as
+			// necessary once the client fix has been released and widely
+			// available for some time.
+			MaxConnectionIdle: 10 * time.Second,
+		}),
+	)
+	joinServiceServer := joinserver.NewJoinServiceGRPCServer(conn.Client)
+	proto.RegisterJoinServiceServer(server, joinServiceServer)
+	process.RegisterCriticalFunc("proxy.grpc.insecure", func() error {
+		log.Infof("Starting proxy gRPC server on %v.", listener.Addr())
+		return trace.Wrap(server.Serve(listener))
+	})
+	return server
+}
+
+// createSecureGRPCServer creates and registers a gRPC server that uses mTLS for
+// authentication. This is used for the gRPC Kube service, which allows users to
+// safely access Kubernetes clusters resources via Teleport without leaking certificates.
+// The gRPC server handles the mTLS because we require the client certificate to be
+// subject in order to determine his identity.
+func createSecureGRPCServer(
+	process *TeleportProcess,
+	log logrus.FieldLogger,
+	limiter *limiter.Limiter,
+	conn *Connector,
+	listener net.Listener,
+	serverTLSConfig *tls.Config,
+) *grpc.Server {
+	if !process.Config.Proxy.Kube.Enabled {
+		return nil
+	}
+	// authMiddleware authenticates request assuming TLS client authentication
+	// adds authentication information to the context
+	// and passes it to the API server
+	authMiddleware := &auth.Middleware{
+		AccessPoint: conn.Client,
+		Limiter:     limiter,
+	}
+
+	server := grpc.NewServer(
+		grpc.ChainUnaryInterceptor(
+			utils.GRPCServerUnaryErrorInterceptor,
+			otelgrpc.UnaryServerInterceptor(),
+			authMiddleware.UnaryInterceptor(),
+		),
+		grpc.ChainStreamInterceptor(
+			utils.GRPCServerStreamErrorInterceptor,
+			otelgrpc.StreamServerInterceptor(),
+			authMiddleware.StreamInterceptor(),
+		),
+		grpc.Creds(credentials.NewTLS(
+			copyAndConfigureTLS(serverTLSConfig, log, authMiddleware.AccessPoint),
+		)),
+	)
+
+	process.RegisterCriticalFunc("proxy.grpc.secure", func() error {
+		log.Infof("Starting proxy gRPC server on %v.", listener.Addr())
+		return trace.Wrap(server.Serve(listener))
+	})
+	return server
+}
+
+// getConfigForClient returns the list of CAs that could have signed the
+// client's certificate.
+func getConfigForClient(log logrus.FieldLogger, accessPoint auth.AccessCache, config *tls.Config) func(info *tls.ClientHelloInfo) (*tls.Config, error) {
```

Comment:
I feel like we have this logic defined in multiple places.  Can we reuse the [WithClusterCAs](https://github.com/gravitational/teleport/blob/e89b9b88933a22962d8472e595c4fe14a3ef6eaa/lib/auth/auth.go#L4527) ?

Reply:
Sure, I did't realize it was already in-place.